src/cargo/sources/registry/http_remote.rs

```diff
@@ -269,6 +297,50 @@ impl<'cfg> HttpRegistry<'cfg> {
             false
         }
     }
+
+    fn config_internal(&mut self) -> Poll<CargoResult<&RegistryConfig>> {
+        if self.registry_config.is_some() {
+            return Poll::Ready(Ok(self.registry_config.as_ref().unwrap()));
+        }
+        debug!("loading config");
+        let index_path = self.assert_index_locked(&self.index_path);
+        let config_json_path = index_path.join("config.json");
+        if self.is_fresh(Path::new("config.json")) {
+            match fs::read(&config_json_path) {
+                Ok(raw_data) => match serde_json::from_slice(&raw_data) {
+                    Ok(json) => {
+                        self.registry_config = Some(json);
+                        return Poll::Ready(Ok(self.registry_config.as_ref().unwrap()));
+                    }
+                    Err(e) => log::debug!("failed to decode cached config.json: {}", e),
+                },
+                Err(e) => log::debug!("failed to read config.json cache: {}", e),
+            }
+        }
+
+        match self.load(Path::new(""), Path::new("config.json"), None)? {
+            Poll::Ready(LoadResponse::Data {
+                raw_data,
+                index_version: _,
+            }) => {
+                trace!("config loaded");
+                self.registry_config = Some(serde_json::from_slice(&raw_data)?);
+                if paths::create_dir_all(&config_json_path.parent().unwrap()).is_ok() {
+                    if let Err(e) = fs::write(&config_json_path, &raw_data) {
+                        log::debug!("failed to write config.json cache: {}", e);
+                    }
+                }
+                Poll::Ready(Ok(self.registry_config.as_ref().unwrap()))
+            }
+            Poll::Ready(LoadResponse::NotFound) => {
+                Poll::Ready(Err(anyhow::anyhow!("config.json not found in registry")))
+            }
+            Poll::Ready(LoadResponse::CacheValid) => {
+                panic!("config.json is not stored in the index cache")
```

Comment:
Can this use `cargo::util::internal(…)` instead? That way, the user gets a nicer message asking them to file a bug report.